packages/flutter_tools/test/general.shard/asset_bundle_variant_test.dart

```diff
@@ -9,84 +9,146 @@ import 'package:file/memory.dart';
+
+  Future<Map<String, List<String>>> extractAssetManifestFromBundle(ManifestAssetBundle bundle) async {
+    final String manifestJson = utf8.decode(await bundle.entries['AssetManifest.json']!.contentsAsBytes());
+    final Map<String, dynamic> parsedJson = json.decode(manifestJson) as Map<String, dynamic>;
+    final Iterable<String> keys = parsedJson.keys;
+    final Map<String, List<String>> parsedManifest = <String, List<String>> {
+      for (final String key in keys) key: List<String>.from(parsedJson[key] as List<dynamic>),
+    };
+    return parsedManifest;
   }
 
   group('AssetBundle asset variants', () {
-    late FileSystem testFileSystem;
-    setUp(() async {
-      testFileSystem = MemoryFileSystem(
-        style: globals.platform.isWindows
-          ? FileSystemStyle.windows
-          : FileSystemStyle.posix,
-      );
-      testFileSystem.currentDirectory = testFileSystem.systemTempDirectory.createTempSync('flutter_asset_bundle_variant_test.');
-    });
+    testWithoutContext('Only images in folders named with device pixel ratios (e.g. 2x, 3.0x) should be considered as variants of other images', () async {
+      final MemoryFileSystem fs = MemoryFileSystem.test();
+
+      const String image = 'assets/image.jpg';
+      const String image2xVariant = 'assets/2x/image.jpg';
+      const String imageNonVariant = 'assets/notAVariant/image.jpg';
+      final List<String> assets = <String>[
+        image,
+        image2xVariant,
+        imageNonVariant
+      ];
 
-    testUsingContext('main asset and variants', () async {
-      globals.fs.file('pubspec.yaml')
-        ..createSync()
-        ..writeAsStringSync(
+      for (final String asset in assets) {
+        final File assetFile = fs.file(asset);
+        await assetFile.create(recursive: true);
+        await assetFile.writeAsString(asset);
+      }
+
+      fs.file('.packages').createSync();
+
+      fs.file('pubspec.yaml').writeAsStringSync(
 '''
 name: test
 dependencies:
   flutter:
     sdk: flutter
 flutter:
   assets:
-    - a/b/c/foo
+    - assets/
 '''
       );
-      globals.fs.file('.packages').createSync();
+
+      final Platform platform = FakePlatform();
+
+      Cache.flutterRoot = Cache.defaultFlutterRoot(
+        platform: platform,
+        fileSystem: fs,
+        userMessages: UserMessages(),
+      );
+
+      final ManifestAssetBundle bundle = ManifestAssetBundle(
+        logger: BufferLogger.test(),
+        fileSystem: fs,
+        platform: platform,
+      );
+
+      await bundle.build(
+        packagesPath: '.packages',
+        assetDirPath: './build/flutter_assets',
+        flutterProject:  FlutterProject.fromDirectoryTest(fs.currentDirectory),
+      );
+
+      final Map<String, List<String>> manifest = await extractAssetManifestFromBundle(bundle);
+      final List<String> variantsForImage = manifest[image]!;
+
+      expect(variantsForImage.contains(image2xVariant), true);
```

Comment:
By using the `contains()` matcher, we'll get a nicer test failure message if it ever fails:

```suggestion
      expect(variantsForImage, contains(image2xVariant));
```

Reply:
Oh nice! I didn't realize we had a matcher library available to us